sdk/spring/azure-identity-spring-library/src/main/java/com/microsoft/azure/identity/spring/SpringEnvironmentTokenBuilder.java

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.microsoft.azure.identity.spring;
+
+import java.util.HashMap;
+import java.util.Optional;
+
+import org.springframework.core.env.Environment;
+
+import com.azure.core.credential.TokenCredential;
+import com.azure.identity.ChainedTokenCredential;
+import com.azure.identity.ChainedTokenCredentialBuilder;
+import com.azure.identity.ClientCertificateCredentialBuilder;
+import com.azure.identity.ClientSecretCredentialBuilder;
+import com.azure.identity.DefaultAzureCredentialBuilder;
+
+/**
+ * A helper class to deal with credentials in a Spring environment.
+ * 
+ * <p>
+ * This helper class makes it possible to configure credentials to be used
+ * within a Spring context.
+ * </p>
+ * 
+ * <table>
+ * <tr>
+ * <th>Property Tuples</th>
+ * <th>Description</th>
+ * </tr>
+ * <tr>
+ * <td>azure.credential.(name.)tenantId <br>
+ * azure.credential.(name.)clientId <br>
+ * azure.credential.(name.)clientSecret</td>
+ * <td>the Azure Tenant ID <br>
+ * the Client ID <br>
+ * the Client Certificate <br>
+ * </td>
+ * </tr>
+ * <tr>
+ * <td>azure.credential.(name.)tenantId <br>
+ * azure.credential.(name.)clientId <br>
+ * azure.credential.(name.)clientCertificate</td>
+ * <td>the Azure Tenant ID <br>
+ * the Client ID <br>
+ * the path to the PEM client certificate</td>
+ * </tr>
+ * </table>
+ * 
+ * where name is the <code>name</code> of the credential. Note if
+ * <code>name</code> is entirely omitted it is taken to be the default
+ * credential. Note if the default credential is omitted it is configure to use
+ * AzureDefaultCredential which allows for the use a Managed Identity (if it is
+ * present).
+ *
+ * @author manfred.riem@microsoft.com
+ */
+public class SpringEnvironmentTokenBuilder {
+
+    /**
+     * Defines the AZURE_CREDENTIAL_PREFIX.
+     */
+    private static final String AZURE_CREDENTIAL_PREFIX = "azure.credential.";
+
+    /**
+     * Stores the named credentials.
+     */
+    private final HashMap<String, TokenCredential> credentials;
+
+    /**
+     * Stores the name of the credential to be returned. If omitted, the default
+     * credential will be returned.
+     */
+    private String name = "";
+
+    /**
+     * Constructor.
+     */
+    public SpringEnvironmentTokenBuilder() {
+        credentials = new HashMap<>();
+        credentials.put("", new DefaultAzureCredentialBuilder().build());
+    }
+
+    /**
+     * Populate from Environment.
+     *
+     * @param environment the environment.
+     */
+    public SpringEnvironmentTokenBuilder fromEnvironment(Environment environment) {
+        populateNamedCredential(environment, "");
+        String credentialNamesKey = AZURE_CREDENTIAL_PREFIX + "names";
```

Comment:
This is a repackaging of @mnriem's code, so perhaps he can answer better. My understanding is that a custom map is needed because property values could come from multiple sources, and this serves to consolidate them in a single place.

Reply:
The 'names' defines the order of the chain.